File: examples/python/flight_simulation.py
#!/usr/bin/env python3
"""
Flight Simulation Example using JaguarEngine Python Bindings

This example demonstrates:
1. Creating an engine and initializing simulation
2. Creating aircraft entities with different domains
3. Setting initial states (position, velocity, orientation)
4. Running the simulation and monitoring state changes
5. Using the event system for collision and threshold monitoring
6. NumPy integration for efficient data handling

Requirements:
    pip install numpy matplotlib

Build JaguarEngine with Python bindings:
    cmake -B build -DJAGUAR_BUILD_PYTHON=ON
    cmake --build build
"""

import sys
import math
from pathlib import Path

# Add build directory to path for development
build_path = Path(__file__).parent.parent.parent / "build"
if build_path.exists():
    sys.path.insert(0, str(build_path))

try:
    import pyjaguar as jag
    import numpy as np
    HAS_MATPLOTLIB = False
    try:
        import matplotlib.pyplot as plt
        HAS_MATPLOTLIB = True
    except ImportError:
        print("matplotlib not available, skipping plots")
except ImportError as e:
    print(f"Failed to import pyjaguar: {e}")
    print("Build with: cmake -B build -DJAGUAR_BUILD_PYTHON=ON && cmake --build build")
    sys.exit(1)


def main():
    """Run a basic flight simulation demonstration."""
    print("=" * 60)
    print("JaguarEngine Python Flight Simulation Example")
    print(f"Version: {jag.__version__}")
    print("=" * 60)

    # ========================================================================
    # 1. Create and Initialize Engine
    # ========================================================================
    print("\n[1] Creating and initializing engine...")

    engine = jag.Engine()
    if not engine.initialize():
        print("Failed to initialize engine!")
        return

    print(f"    Engine state: {engine.get_state()}")
    print(f"    Initial time: {engine.get_time():.3f} s")

    # ========================================================================
    # 2. Create Aircraft Entity
    # ========================================================================
    print("\n[2] Creating aircraft entity...")

    aircraft_id = engine.create_entity("F16_Alpha", jag.Domain.Air)
    if aircraft_id == jag.INVALID_ENTITY_ID:
        print("Failed to create aircraft!")
        return

    print(f"    Created aircraft with ID: {aircraft_id}")

    # ========================================================================
    # 3. Set Initial State
    # ========================================================================
    print("\n[3] Setting initial state...")

    initial_state = jag.EntityState()

    # Position: 10km altitude, at origin
    initial_state.position = jag.Vec3(0.0, 0.0, -10000.0)  # NED: down is negative

    # Velocity: 200 m/s forward (approximately Mach 0.6)
    initial_state.velocity = jag.Vec3(200.0, 0.0, 0.0)

    # Orientation: Level flight, heading north
    initial_state.orientation = jag.Quat.identity()

    # Mass: typical F-16 empty weight
    initial_state.mass = 8500.0  # kg

    # Inertia tensor (simplified)
    initial_state.inertia = jag.Mat3x3.inertia(
        ixx=12000.0,  # Roll
        iyy=75000.0,  # Pitch
        izz=85000.0   # Yaw
    )

    engine.set_entity_state(aircraft_id, initial_state)

    state = engine.get_entity_state(aircraft_id)
    print(f"    Position: ({state.position.x:.1f}, {state.position.y:.1f}, {state.position.z:.1f}) m")
    print(f"    Velocity: ({state.velocity.x:.1f}, {state.velocity.y:.1f}, {state.velocity.z:.1f}) m/s")
    print(f"    Mass: {state.mass:.1f} kg")

    # ========================================================================
    # 4. Set Up Event Monitoring (Optional)
    # ========================================================================
    print("\n[4] Setting up event monitoring...")

    # Event callback function
    event_log = []

    def on_event(event_data):
        """Handle all events."""
        event_log.append({
            'time': event_data.timestamp,
            'type': event_data.type,
            'source': event_data.source_entity
        })
        return True  # Continue propagation

    # Create event dispatcher (if available in engine)
    try:
        dispatcher = jag.EventDispatcher()
        handler_id = dispatcher.subscribe_all(on_event, "flight_monitor")
        print(f"    Event handler registered: {handler_id}")
    except Exception as e:
        print(f"    Event system not available: {e}")
        dispatcher = None

    # ========================================================================
    # 5. Run Simulation
    # ========================================================================
    print("\n[5] Running simulation...")

    dt = 0.01  # 10ms time step (100 Hz)
    sim_duration = 10.0  # 10 seconds
    num_steps = int(sim_duration / dt)

    # Data recording
    times = []
    positions_x = []
    positions_z = []
    velocities = []

    print(f"    Time step: {dt * 1000:.1f} ms")
    print(f"    Duration: {sim_duration:.1f} s")
    print(f"    Total steps: {num_steps}")
    print()

    for i in range(num_steps):
        # Step simulation
        engine.step(dt)

        # Record state every 100 steps (1 second intervals)
        if i % 100 == 0:
            state = engine.get_entity_state(aircraft_id)
            t = engine.get_time()

            times.append(t)
            positions_x.append(state.position.x)
            positions_z.append(-state.position.z)  # Convert to altitude
            velocities.append(state.velocity.length())

            # Print progress
            print(f"    t={t:6.2f}s | "
                  f"x={state.position.x:10.1f}m | "
                  f"alt={-state.position.z:8.1f}m | "
                  f"v={state.velocity.length():6.1f}m/s")

    # ========================================================================
    # 6. Display Results
    # ========================================================================
    print("\n[6] Final state:")

    final_state = engine.get_entity_state(aircraft_id)
    print(f"    Position: ({final_state.position.x:.1f}, {final_state.position.y:.1f}, {final_state.position.z:.1f}) m")
    print(f"    Velocity: ({final_state.velocity.x:.1f}, {final_state.velocity.y:.1f}, {final_state.velocity.z:.1f}) m/s")
    print(f"    Distance traveled: {final_state.position.x:.1f} m")
    print(f"    Simulation time: {engine.get_time():.3f} s")

    if event_log:
        print(f"\n    Events logged: {len(event_log)}")

    # ========================================================================
    # 7. Plot Results (if matplotlib available)
    # ========================================================================
    if HAS_MATPLOTLIB and len(times) > 1:
        print("\n[7] Generating plots...")

        fig, axes = plt.subplots(2, 2, figsize=(12, 8))
        fig.suptitle("JaguarEngine Flight Simulation Results", fontsize=14)

        # Position over time
        axes[0, 0].plot(times, np.array(positions_x) / 1000, 'b-')
        axes[0, 0].set_xlabel('Time (s)')
        axes[0, 0].set_ylabel('Distance (km)')
        axes[0, 0].set_title('Horizontal Distance')
        axes[0, 0].grid(True)

        # Altitude over time
        axes[0, 1].plot(times, np.array(positions_z) / 1000, 'g-')
        axes[0, 1].set_xlabel('Time (s)')
        axes[0, 1].set_ylabel('Altitude (km)')
        axes[0, 1].set_title('Altitude')
        axes[0, 1].grid(True)

        # Velocity over time
        axes[1, 0].plot(times, velocities, 'r-')
        axes[1, 0].set_xlabel('Time (s)')
        axes[1, 0].set_ylabel('Speed (m/s)')
        axes[1, 0].set_title('Speed')
        axes[1, 0].grid(True)

        # Trajectory (x vs altitude)
        axes[1, 1].plot(np.array(positions_x) / 1000, np.array(positions_z) / 1000, 'k-')
        axes[1, 1].set_xlabel('Distance (km)')
        axes[1, 1].set_ylabel('Altitude (km)')
        axes[1, 1].set_title('Trajectory')
        axes[1, 1].grid(True)

        plt.tight_layout()
        plt.savefig('flight_simulation_results.png', dpi=150)
        print("    Saved: flight_simulation_results.png")
        plt.show()

    # ========================================================================
    # 8. Cleanup
    # ========================================================================
    print("\n[8] Shutting down...")

    engine.destroy_entity(aircraft_id)
    engine.shutdown()

    print("    Done!")
    print("=" * 60)


def demo_numpy_integration():
    """Demonstrate NumPy integration for batch operations."""
    print("\n" + "=" * 60)
    print("NumPy Integration Demo")
    print("=" * 60)

    # Vec3 <-> NumPy conversion
    v = jag.Vec3(1.0, 2.0, 3.0)
    arr = v.to_numpy()
    print(f"\nVec3 to NumPy: {v} -> {arr}")

    # NumPy operations
    arr_normalized = arr / np.linalg.norm(arr)
    v_normalized = jag.Vec3.from_numpy(arr_normalized)
    print(f"Normalized: {v_normalized}")

    # Mat3x3 <-> NumPy conversion
    m = jag.Mat3x3.diagonal(1.0, 2.0, 3.0)
    m_arr = m.to_numpy()
    print(f"\nMat3x3 to NumPy:\n{m_arr}")

    # NumPy matrix operations
    m_inv = np.linalg.inv(m_arr)
    print(f"Inverse:\n{m_inv}")

    # Quat <-> NumPy conversion
    q = jag.Quat.from_euler(0.1, 0.2, 0.3)
    q_arr = q.to_numpy()
    print(f"\nQuat to NumPy (w,x,y,z): {q_arr}")


def demo_threshold_monitoring():
    """Demonstrate threshold monitoring for parameter tracking."""
    print("\n" + "=" * 60)
    print("Threshold Monitoring Demo")
    print("=" * 60)

    # Create threshold monitor
    monitor = jag.ThresholdMonitor()

    # Configure speed threshold
    speed_config = jag.speed_threshold(max_speed=340.0, hysteresis=10.0)  # Mach 1
    monitor.add_threshold("speed", speed_config)

    # Configure altitude threshold
    alt_config = jag.altitude_threshold(max_altitude=15000.0, min_altitude=500.0)
    monitor.add_threshold("altitude", alt_config)

    # Configure fuel threshold
    fuel_config = jag.fuel_threshold(low_warning=20.0, critical_low=5.0)
    monitor.add_threshold("fuel", fuel_config)

    print(f"\nConfigured {monitor.threshold_count()} thresholds")

    # Simulate parameter changes
    test_values = [
        ("speed", 300.0, "Normal speed"),
        ("speed", 350.0, "Overspeed!"),
        ("speed", 320.0, "Still overspeed (hysteresis)"),
        ("speed", 325.0, "Recovered to normal"),
        ("altitude", 10000.0, "Normal altitude"),
        ("altitude", 16000.0, "Altitude exceeded!"),
        ("fuel", 50.0, "Fuel OK"),
        ("fuel", 15.0, "Fuel low warning!"),
        ("fuel", 3.0, "Fuel critical!"),
    ]

    print("\nSimulating parameter changes:")
    for param, value, description in test_values:
        monitor.update(param, value)
        state = monitor.get_state(param)
        print(f"  {param}={value:6.1f} -> {state.name:15s} | {description}")

    # Check exceeded thresholds
    exceeded = monitor.get_exceeded_thresholds()
    if exceeded:
        print(f"\nExceeded thresholds: {exceeded}")

    print()


if __name__ == "__main__":
    main()
    demo_numpy_integration()
    demo_threshold_monitoring()

File: examples/lua/flight_simulation.lua
#!/usr/bin/env lua
--[[
    Flight Simulation Example using JaguarEngine Lua Bindings

    This example demonstrates:
    1. Creating an engine and initializing simulation
    2. Creating aircraft entities with different domains
    3. Setting initial states (position, velocity, orientation)
    4. Running the simulation and monitoring state changes
    5. Using the event system for collision and threshold monitoring

    Build JaguarEngine with Lua bindings:
        cmake -B build -DJAGUAR_BUILD_LUA=ON
        cmake --build build
]]

-- Add build directory to package.path
package.cpath = package.cpath .. ";./build/?.so;./build/?.dll"

-- Try to load JaguarEngine
local status, jag = pcall(require, "jaguar")
if not status then
    print("Failed to load jaguar module: " .. tostring(jag))
    print("Build with: cmake -B build -DJAGUAR_BUILD_LUA=ON && cmake --build build")
    os.exit(1)
end

-- Print banner
print(string.rep("=", 60))
print("JaguarEngine Lua Flight Simulation Example")
print(string.format("Version: %s", JAGUAR_VERSION))
print(string.rep("=", 60))

-- ========================================================================
-- 1. Create and Initialize Engine
-- ========================================================================
print("\n[1] Creating and initializing engine...")

local engine = Engine()
if not engine:initialize() then
    print("Failed to initialize engine!")
    os.exit(1)
end

print(string.format("    Engine state: %s", tostring(engine:get_state())))
print(string.format("    Initial time: %.3f s", engine:get_time()))

-- ========================================================================
-- 2. Create Aircraft Entity
-- ========================================================================
print("\n[2] Creating aircraft entity...")

local aircraft_id = engine:create_entity("F16_Alpha", Domain.Air)
if aircraft_id == 0 then  -- INVALID_ENTITY_ID
    print("Failed to create aircraft!")
    os.exit(1)
end

print(string.format("    Created aircraft with ID: %d", aircraft_id))

-- ========================================================================
-- 3. Set Initial State
-- ========================================================================
print("\n[3] Setting initial state...")

local initial_state = EntityState()

-- Position: 10km altitude, at origin (NED: down is negative)
initial_state.position = Vec3(0.0, 0.0, -10000.0)

-- Velocity: 200 m/s forward (approximately Mach 0.6)
initial_state.velocity = Vec3(200.0, 0.0, 0.0)

-- Orientation: Level flight, heading north
initial_state.orientation = Quat.identity()

-- Mass: typical F-16 empty weight
initial_state.mass = 8500.0  -- kg

-- Inertia tensor (simplified)
initial_state.inertia = Mat3x3.inertia(12000.0, 75000.0, 85000.0)

engine:set_entity_state(aircraft_id, initial_state)

local state = engine:get_entity_state(aircraft_id)
print(string.format("    Position: (%.1f, %.1f, %.1f) m",
    state.position.x, state.position.y, state.position.z))
print(string.format("    Velocity: (%.1f, %.1f, %.1f) m/s",
    state.velocity.x, state.velocity.y, state.velocity.z))
print(string.format("    Mass: %.1f kg", state.mass))

-- ========================================================================
-- 4. Set Up Event Monitoring (Optional)
-- ========================================================================
print("\n[4] Setting up event monitoring...")

-- Event log table
local event_log = {}

-- Create event dispatcher
local dispatcher = EventDispatcher()

-- Subscribe to all events
local handler_id = dispatcher:subscribe_all(function(event_data)
    table.insert(event_log, {
        time = event_data.timestamp,
        type = event_data.type,
        source = event_data.source_entity
    })
    return true  -- Continue propagation
end, "flight_monitor", 0)

print(string.format("    Event handler registered: %d", handler_id))

-- ========================================================================
-- 5. Run Simulation
-- ========================================================================
print("\n[5] Running simulation...")

local dt = 0.01  -- 10ms time step (100 Hz)
local sim_duration = 10.0  -- 10 seconds
local num_steps = math.floor(sim_duration / dt)

-- Data recording
local times = {}
local positions_x = {}
local positions_z = {}
local velocities = {}

print(string.format("    Time step: %.1f ms", dt * 1000))
print(string.format("    Duration: %.1f s", sim_duration))
print(string.format("    Total steps: %d", num_steps))
print()

for i = 1, num_steps do
    -- Step simulation
    engine:step(dt)

    -- Record state every 100 steps (1 second intervals)
    if i % 100 == 0 then
        local state = engine:get_entity_state(aircraft_id)
        local t = engine:get_time()

        table.insert(times, t)
        table.insert(positions_x, state.position.x)
        table.insert(positions_z, -state.position.z)  -- Convert to altitude
        table.insert(velocities, state.velocity:length())

        -- Print progress
        print(string.format("    t=%6.2fs | x=%10.1fm | alt=%8.1fm | v=%6.1fm/s",
            t, state.position.x, -state.position.z, state.velocity:length()))
    end
end

-- ========================================================================
-- 6. Display Results
-- ========================================================================
print("\n[6] Final state:")

local final_state = engine:get_entity_state(aircraft_id)
print(string.format("    Position: (%.1f, %.1f, %.1f) m",
    final_state.position.x, final_state.position.y, final_state.position.z))
print(string.format("    Velocity: (%.1f, %.1f, %.1f) m/s",
    final_state.velocity.x, final_state.velocity.y, final_state.velocity.z))
print(string.format("    Distance traveled: %.1f m", final_state.position.x))
print(string.format("    Simulation time: %.3f s", engine:get_time()))

if #event_log > 0 then
    print(string.format("\n    Events logged: %d", #event_log))
end

-- ========================================================================
-- 7. Cleanup
-- ========================================================================
print("\n[7] Shutting down...")

engine:remove_entity(aircraft_id)
engine:shutdown()

print("    Done!")
print(string.rep("=", 60))

-- ========================================================================
-- Bonus: Demonstrate Core Types
-- ========================================================================
print("\n" .. string.rep("=", 60))
print("Core Types Demonstration")
print(string.rep("=", 60))

-- Vec3 operations
print("\nVec3 Operations:")
local v1 = Vec3(1, 2, 3)
local v2 = Vec3(4, 5, 6)
print(string.format("  v1 = %s", tostring(v1)))
print(string.format("  v2 = %s", tostring(v2)))
print(string.format("  v1 + v2 = %s", tostring(v1 + v2)))
print(string.format("  v1 . v2 = %.2f (dot)", v1:dot(v2)))
print(string.format("  v1 x v2 = %s (cross)", tostring(v1:cross(v2))))
print(string.format("  |v1| = %.4f", v1:length()))
print(string.format("  normalized = %s", tostring(v1:normalized())))

-- Quaternion operations
print("\nQuaternion Operations:")
local q = Quat.from_euler(0.1, 0.2, 0.3)
print(string.format("  q from euler(0.1, 0.2, 0.3) = %s", tostring(q)))
local roll, pitch, yaw = q:to_euler()
print(string.format("  back to euler: roll=%.3f, pitch=%.3f, yaw=%.3f", roll, pitch, yaw))

-- Rotate a vector
local v = Vec3(1, 0, 0)
local rotated = q:rotate(v)
print(string.format("  rotate Vec3(1,0,0) = %s", tostring(rotated)))

-- Mat3x3 operations
print("\nMat3x3 Operations:")
local m = Mat3x3.identity()
print(string.format("  Identity matrix trace: %.1f", m:trace()))
local inertia = Mat3x3.inertia(100, 200, 300)
print(string.format("  Inertia matrix determinant: %.1f", inertia:determinant()))

-- Constants
print("\nPhysical Constants:")
print(string.format("  PI = %.10f", constants.PI))
print(string.format("  G0 (standard gravity) = %.5f m/s^2", constants.G0))
print(string.format("  Speed of light = %.0f m/s", constants.C))
print(string.format("  Earth radius (equator) = %.1f m", constants.R_EARTH_EQUATOR))

-- Unit conversions
print("\nUnit Conversions:")
print(string.format("  180 deg = %.6f rad", deg_to_rad(180)))
print(string.format("  PI rad = %.1f deg", rad_to_deg(constants.PI)))
print(string.format("  1 ft = %.4f m", ft_to_m(1)))
print(string.format("  100 kt = %.2f m/s", kt_to_ms(100)))

print("\n" .. string.rep("=", 60))
print("All Lua bindings demonstrated successfully!")
print(string.rep("=", 60))
